include/component.hxx

```diff
@@ -10,43 +10,159 @@
 #include <string>
 #include <memory>
 
+#include "guarded_options.hxx"
+#include "permissions.hxx"
+#include "hermes_utils.hxx"
+
 class Solver; // Time integrator
 
+/// Simple struct to store information on the different types of
+/// species present in a simulation
+struct SpeciesInformation {
+  SpeciesInformation(const std::vector<std::string>& electrons,
+                     const std::vector<std::string>& neutrals,
+                     const std::vector<std::string>& positive_ions,
+                     const std::vector<std::string> & negative_ions)
+    : electrons(electrons), neutrals(neutrals), positive_ions(positive_ions), negative_ions(negative_ions), ions(positive_ions) {
+    finish_construction();
+  }
+
+  SpeciesInformation(const std::initializer_list<std::string> species) {
+    for (auto& sp : species) {
+      // FIXME: identifySpecies only identifies positive ions
+      // FIXME: identifySpecies has no concept of ebeam
+      SpeciesType type = identifySpeciesType(sp);
+      if (type == SpeciesType::electron) {
+        electrons.push_back(sp);
+      } else if (type == SpeciesType::ion) {
+        positive_ions.push_back(sp);
+      } else if (type == SpeciesType::neutral) {
+        neutrals.push_back(sp);
+      } else {
+        throw BoutException("Species {} has unrecognised type {}", sp, toString(type));
+      }
+      finish_construction();
+    }
+  }
+
+  std::vector<std::string> electrons, neutrals, positive_ions, negative_ions, ions, charged, non_electrons, all_species;
+
+  private:
+    void finish_construction() {
+      ions = positive_ions;
+      ions.insert(ions.end(), negative_ions.begin(), negative_ions.end());
+      charged = ions;
+      charged.insert(charged.end(), electrons.begin(), electrons.end());
+      non_electrons = ions;
+      non_electrons.insert(non_electrons.end(), neutrals.begin(), neutrals.end());
+      all_species = charged;
+      all_species.insert(all_species.end(), neutrals.begin(), neutrals.end());
+    }
+};
+
+/// Lightweight type to store information used to build a component.
+struct ComponentInformation {
+  std::string name;
+  std::string type;
+
+  ComponentInformation(const std::string& name_, const std::string& type_) : name(name_), type(type_) {}
+
+  ComponentInformation(std::string&& name_, std::string&& type_) : name(std::move(name_)), type(std::move(type_)) {}
+
```

Comment:
We should be able to rely on aggregate intialisation here, and so no need for constructors:

```suggestion
```